File: bigdata2020f/src/main/java/ca/uwaterloo/cs451/a1/StripesPMI.java
/**
 * Bespin: reference implementations of "big data" algorithms
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ca.uwaterloo.cs451.a1;

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.StringTokenizer;

import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.CommandLineParser;
import org.apache.commons.cli.GnuParser;
import org.apache.commons.cli.HelpFormatter;
import org.apache.commons.cli.OptionBuilder;
import org.apache.commons.cli.Options;
import org.apache.commons.cli.ParseException;
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.conf.Configured;
import org.apache.hadoop.fs.FSDataInputStream;
import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.io.DoubleWritable;
import org.apache.hadoop.io.IntWritable;
import org.apache.hadoop.io.LongWritable;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mapreduce.Job;
import org.apache.hadoop.mapreduce.Mapper;
import org.apache.hadoop.mapreduce.Reducer;
import org.apache.hadoop.mapreduce.lib.input.FileInputFormat;
import org.apache.hadoop.mapreduce.lib.output.FileOutputFormat;
import org.apache.hadoop.mapreduce.lib.output.TextOutputFormat;
import org.apache.hadoop.util.Tool;
import org.apache.hadoop.util.ToolRunner;
import org.apache.log4j.Logger;


import io.bespin.java.util.Tokenizer;
import org.apache.hadoop.conf.Configured;
import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.io.LongWritable;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mapreduce.Job;
import org.apache.hadoop.mapreduce.Mapper;
import org.apache.hadoop.mapreduce.Reducer;
import org.apache.hadoop.mapreduce.lib.input.FileInputFormat;
import org.apache.hadoop.mapreduce.lib.output.FileOutputFormat;
import org.apache.hadoop.util.Tool;
import org.apache.hadoop.util.ToolRunner;
import org.apache.log4j.Logger;
import org.kohsuke.args4j.CmdLineException;
import org.kohsuke.args4j.CmdLineParser;
import org.kohsuke.args4j.Option;
import org.kohsuke.args4j.ParserProperties;
import tl.lin.data.map.HMapStIW;
import tl.lin.data.pair.*;
import org.apache.hadoop.io.IntWritable;
import java.io.IOException;
import java.util.Iterator;
import java.util.List;
import java.util.*;
import java.lang.*;


public class StripesPMI extends Configured implements Tool {

  private static final Logger LOG = Logger.getLogger(StripesPMI.class);

  // First stage Mapper: emits pairs (token, 1) for each unique token in an input value,
  //                  and a pair (token pair, 1) for each unique pair of tokens
      private static class AppearanceCountMapper extends Mapper<LongWritable, Text, Text, IntWritable> {

    // Objects for reuse
    private final static Text KEY = new Text();
    private final static IntWritable ONE = new IntWritable(1);

    @Override
    public void map(LongWritable key, Text value, Context context)
        throws IOException, InterruptedException{

      String line = value.toString();
      StringTokenizer t = new StringTokenizer(line);
      Set<String> unique = new HashSet<String>();
      String token = "";

      while(t.hasMoreTokens()){
        token = t.nextToken();

        if(unique.add(token)){
          KEY.set(token);
          context.write(KEY, ONE);
        }
      }

    }
  }

  // First stage Reducer: Totals counts for each Token
  private static class AppearanceCountReducer extends Reducer<Text, IntWritable, Text, IntWritable> {
    // Reuse objects
    private final static IntWritable SUM = new IntWritable();

    @Override
    public void reduce(Text key, Iterable<IntWritable> values, Context context)
        throws IOException, InterruptedException{
      int sum = 0;
      for(IntWritable value : values){
        sum += value.get();
      }

      SUM.set(sum);
      context.write(key, SUM);
    }
  }



  // Second stage mapper: Maps key: term, value: map term to "cooccurance" neighbors 
  private static class StripesPMIMapper extends Mapper<LongWritable, Text, Text, HMapStIW> {

    // Objects for reuse
    private final static Text KEY = new Text();
    private final static HMapStIW MAP = new HMapStIW();

    @Override
    public void map(LongWritable key, Text value, Context context) 
        throws IOException, InterruptedException{

      String line = value.toString();
      StringTokenizer t = new StringTokenizer(line);

      // Need to pass through multiple times so put tokens into array
      List<String> terms = new ArrayList<String>();
      while(t.hasMoreTokens()){
        terms.add(t.nextToken());
      }


      String left = "";
      String right = "";
      for(int leftTermIndex = 0; leftTermIndex < terms.size(); leftTermIndex++){
        left = terms.get(leftTermIndex);

        for(int rightTermIndex = leftTermIndex + 1; rightTermIndex < terms.size(); rightTermIndex++) {
          right = terms.get(rightTermIndex);

          if(!MAP.containsKey(right)){
            MAP.put(right, 1);
          }

        } // Each right word put in map


        KEY.set(left);
        context.write(KEY, MAP);
        MAP.clear();
      }


    }
  }

  //TODO Write fill in combiner boilerplate
  //Combiner
  private static class StripesPMICombiner extends Reducer<Text, HMapStIW, Text, HMapStIW> {
    private static HMapStIW MAP = new HMapStIW(); 



    @Override
    public void reduce(Text term, Iterable<HMapStIW> values, Context context) 
        throws IOException, InterruptedException{

      // Do a element-wise sum of the maps
      for(HMapStIW pairMap : values){

        for(String key : pairMap.keySet()){

          if(MAP.containsKey(key)){
            MAP.put(key, MAP.get(key) + pairMap.get(key));
          }else{
            MAP.put(key, pairMap.get(key));
          }

        }

      }
      context.write(term, MAP);
    }
  }

  // Second Stage reducer: Finalizes PMI Calculation given 
  private static class StripesPMIReducer extends Reducer<Text, HMapStIW, PairOfStrings, PairOfWritables<DoubleWritable,IntWritable>> {

    private static Map<String, Integer> termTotals = new HashMap<String, Integer>();
    private static Map<String, Integer> MAP = new HashMap<String, Integer>();
    private static PairOfStrings PAIR = new PairOfStrings();
    private static DoubleWritable PMI = new DoubleWritable();
    private static double totalDocs = 156215;

    @Override
    public void setup(Context context) throws IOException{
      //TODO Read from intermediate output of first job
      // and build in-memory map of terms to their individual totals
//       final Args args = new Args();
      Configuration conf = context.getConfiguration();
      FileSystem fs = FileSystem.get(conf);

      //      Path inFile = new Path(conf.get("intermediatePath"));
      Path inFile = new Path("cs451-bigdatateach-inter/part-r-00000");

      if(!fs.exists(inFile)){
        throw new IOException("File Not Found: " + inFile.toString());
      }

      BufferedReader reader = null;
      try{
        FSDataInputStream in = fs.open(inFile);
        InputStreamReader inStream = new InputStreamReader(in);
        reader = new BufferedReader(inStream);

      } catch(FileNotFoundException e){
        throw new IOException("Exception thrown when trying to open file.");
      }


      String line = reader.readLine();
      while(line != null){

        String[] parts = line.split("\\s+");
        if(parts.length != 2){
          LOG.info("Input line did not have exactly 2 tokens: '" + line + "'");
        } else {
          termTotals.put(parts[0], Integer.parseInt(parts[1]));
        }
        line = reader.readLine();
      }

      reader.close();

    }

    @Override
    public void reduce(Text term, Iterable<HMapStIW> values, Context context ) 
        throws IOException, InterruptedException{
      // Recieving pair and pair counts -> Sum these for this pair's total
      // Only calculate PMI for pairs that occur 10 or more times


      // Do a element-wise sum of the maps
      for(HMapStIW pairMap : values){

        for(String key : pairMap.keySet()){

          if(MAP.containsKey(key)){
            MAP.put(key, MAP.get(key) + pairMap.get(key));
          }else{
            MAP.put(key, pairMap.get(key));
          }
        }
      }

      // MAP contians the total co-appearnaces for incoming key, "term" 
      //    and all the keys of the map. We'll make Pairs like this: (term, key_i)
      String leftTerm = term.toString();

      for(String key : MAP.keySet()){
        PAIR.set(leftTerm, key);

        double probPair = MAP.get(key) / totalDocs;
        double probLeft = termTotals.get(leftTerm) / totalDocs;
        double probRight = termTotals.get(key) / totalDocs;

        double pmi = Math.log(probPair / (probLeft * probRight));

        PMI.set(pmi);
	         IntWritable ans=new IntWritable(MAP.get(key));
        context.write(PAIR, new PairOfWritables<>(PMI,ans));
      }
      
      MAP.clear();

    }
    
  }



 private static final class Args {
    @Option(name = "-input", metaVar = "[path]", required = true, usage = "input path")
    String input;

    @Option(name = "-output", metaVar = "[path]", required = true, usage = "output path")
    String output;

    @Option(name = "-reducers", metaVar = "[num]", usage = "number of reducers")
    int numReducers = 1;

    @Option(name = "-window", metaVar = "[num]", usage = "cooccurrence window")
    int window = 2;
  }


  public StripesPMI() {}

  private static final String INPUT = "input";
  private static final String OUTPUT = "output";
  private static final String NUM_REDUCERS = "numReducers";

  @SuppressWarnings("static-access")
  public int run(String[] argv) throws Exception {
//    Options options = new Options();
	final Args args = new Args();
  //  options.addOption(OptionBuilder.withArgName("path").hasArg()
    //    .withDescription("input path").create(INPUT));
    //options.addOption(OptionBuilder.withArgName("path").hasArg()
     //   .withDescription("output path").create(OUTPUT));
    //options.addOption(OptionBuilder.withArgName("num").hasArg()
      //  .withDescription("number of reducers").create(NUM_REDUCERS));

  //  CommandLine cmdline;
   // CommandLineParser parser = new GnuParser();

//    try{
  //    cmdline = parser.parse(options, args);
   // } catch (ParseException exp) {
     // System.err.println("Error parsing command line: " + exp.getMessage());
      //return -1;
    //}

    /*if(!cmdline.hasOption(INPUT) || !cmdline.hasOption(OUTPUT)) {
      System.out.println("args: " + Arrays.toString(args));
      HelpFormatter formatter = new HelpFormatter();
      formatter.setWidth(120);
      formatter.printHelp(this.getClass().getName(), options);
      ToolRunner.printGenericCommandUsage(System.out);
      return -1;
    } */

//    String inputPath = cmdline.getOptionValue(INPUT);

    //TODO This output path is for the 2nd job's.
    //    The fits job will have an intermediate output path from which the second job's reducer will read
//    String outputPath = cmdline.getOptionValue(OUTPUT);
    String intermediatePath = "cs451-bigdatateach-inter";




//    int reduceTasks = cmdline.hasOption(NUM_REDUCERS) ? 
  //      Integer.parseInt(cmdline.getOptionValue(NUM_REDUCERS)) : 1;

        LOG.info("Tool: " + StripesPMI.class.getSimpleName() + " Appearances Count");
        LOG.info(" - input path: " + args.input);
        LOG.info(" - output path: " + intermediatePath);
        LOG.info(" - number of reducers: " + args.numReducers);

        Configuration conf = getConf();
        conf.set("intermediatePath", intermediatePath);

        Job job1 = Job.getInstance(conf);
        job1.setJobName(StripesPMI.class.getSimpleName() + " Appearance Count");
        job1.setJarByClass(StripesPMI.class);

        job1.setNumReduceTasks(args.numReducers);

        FileInputFormat.setInputPaths(job1, new Path(args.input));
        FileOutputFormat.setOutputPath(job1, new Path(intermediatePath));

        job1.setOutputKeyClass(Text.class);
        job1.setOutputValueClass(IntWritable.class);

        job1.setMapperClass(AppearanceCountMapper.class);
        job1.setCombinerClass(AppearanceCountReducer.class);
        job1.setReducerClass(AppearanceCountReducer.class);

        // Delete the output directory if it exists already.
        Path intermediateDir = new Path(intermediatePath);
        FileSystem.get(conf).delete(intermediateDir, true);

        long startTime = System.currentTimeMillis();
        job1.waitForCompletion(true);
        LOG.info("Job Finished in " + (System.currentTimeMillis() - startTime) / 1000.0 + " seconds");


        // Start second job
        LOG.info("Tool: " + StripesPMI.class.getSimpleName() + " Stripes PMI Part");
        LOG.info(" - input path: " + args.input);
        LOG.info(" - output path: " + args.output);
        LOG.info(" - number of reducers: " + args.numReducers);

        Job job2 = Job.getInstance(conf);
        job2.setJobName(StripesPMI.class.getSimpleName() + " StripesPMICalcuation");
        job2.setJarByClass(StripesPMI.class);

        job2.setNumReduceTasks(args.numReducers);

        FileInputFormat.setInputPaths(job2,  new Path(args.input));
        TextOutputFormat.setOutputPath(job2, new Path(args.output));

        //TODO Which output key??
        job2.setOutputKeyClass(Text.class);
        job2.setOutputValueClass(HMapStIW.class);
        job2.setMapperClass(StripesPMIMapper.class);
        job2.setCombinerClass(StripesPMICombiner.class);
        job2.setReducerClass(StripesPMIReducer.class);

        Path outputDir = new Path(args.output);
        FileSystem.get(conf).delete(outputDir, true);

        startTime = System.currentTimeMillis();
        job2.waitForCompletion(true);
        LOG.info("Job Finished in " + (System.currentTimeMillis() - startTime) / 1000.0 + " seconds");


        return 0;
  }

  /**
   * @param args
   */
  public static void main(String[] args) throws Exception{
    ToolRunner.run(new StripesPMI(), args);
  }
}

File: bigdata2020f/src/main/scala/ca/uwaterloo/cs451/a2/StripesPMI.scala
/**
  * Bespin: reference implementations of "big data" algorithms
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  * http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */

package ca.uwaterloo.cs451.a2

import io.bespin.scala.util.Tokenizer

import org.apache.log4j._
import org.apache.hadoop.fs._
import org.apache.spark.SparkContext
import org.apache.spark.SparkConf
import org.rogach.scallop._
import org.apache.spark.Partitioner



class StripesPMIConf(args: Seq[String]) extends ScallopConf(args) with Tokenizer {
  mainOptions = Seq(input, output, reducers, threshold)
  val input = opt[String](descr = "input path", required = true)
  val output = opt[String](descr = "output path", required = true)
  val reducers = opt[Int](descr = "number of reducers", required = false, default = Some(1))
  val threshold = opt[Int](descr = "threshold", required = false, default = Some(10))
  val numExecutors = opt[Int](descr = "number of executors", required = false, default = Some(1))
  val executorCores = opt[Int](descr = "number of cores", required = false, default = Some(1))
  verify()
}




object StripesPMI extends Tokenizer {
  val log = Logger.getLogger(getClass().getName())

  def main(argv: Array[String]) {
    val args = new StripesPMIConf(argv)

    log.info("Input: " + args.input())
    log.info("Output: " + args.output())
    log.info("Number of reducers: " + args.reducers())

    val conf = new SparkConf().setAppName("StripesPMI")
    val sc = new SparkContext(conf)
    val threshold=args.threshold()    
    val red=args.reducers()    
    val outputDir = new Path(args.output())
    FileSystem.get(sc.hadoopConfiguration).delete(outputDir, true)

    val interDir=new Path("cs451-inter-pairspmi")
    FileSystem.get(sc.hadoopConfiguration).delete(interDir,true)
    val path="cs451-inter-pairspmi"


     var txt=sc.textFile(args.input())
    var total_lines=txt.count()
    var words=txt.flatMap(line=>{
    var tokens=tokenize(line)
    if(tokens.length>0)
    {
	//List()
	tokens.take(Math.min(tokens.length,40)).distinct
//   tokens.distinct
    }
   else List()
    
    
    
    }).map(p=>(p,1)).reduceByKey(_+_).collectAsMap()
    
    var broadcast_var=sc.broadcast(words)
    
    txt.flatMap(line=>{
    var tokens=tokenize(line)
    val words=tokens.take(Math.min(tokens.length,40)).distinct
    var iter=scala.collection.mutable.ListBuffer[(String,Map[String,Int])]()
//    var mapper=scala.collection.mutable.Map[String,Integer]()
    if(words.length>1)
    {
    var i=0
    var j=0
    for(i<- 0 to words.length-1)
    {
    var mapper=scala.collection.mutable.Map[String,Int]()  // for each ith word we create its hashmap
    for(j<- 0 to words.length-1)
    {
    if((i!=j) && (words(i)!=words(j)))
    {
    var value=mapper.getOrElse(words(j),0)+1
    mapper(words(j))=value
    }
    //var pair:(String,String)= (tokens(i),tokens(j))
    //mapper+=tokens(i)->tokens(j)
    }
    var temp:(String,Map[String,Int])=(words(i),mapper.toMap)
    iter+=temp
    
    }
    iter.toList
    }
    
    
    else List()
    
    
    
    
    }).reduceByKey((stripe1,stripe2)=>{ stripe1++stripe2 .map{ case (k,v)=> 
   k->(v+stripe1.getOrElse(k,0))
   
   }
    
    
    
    
    
    },red).map(stripe=>{
    
    var x=stripe._1

	(stripe._1,stripe._2.filter((m)=>m._2>=threshold).map{ case (k,v)=>{

	var y=broadcast_var.value(k)
	var pmi=Math.log10((total_lines.toFloat*v)/(broadcast_var.value(x).toFloat*y))
	//k +  " -> (" + pmi +","+ v + ")"
	k->(pmi,v)


}})
}).filter((m)=>m._2.size>0).map(stripe=>(stripe._1,stripe._2.toMap)).saveAsTextFile(args.output())
//  uncomment this part   map(stripe=> "(" + stripe._1+",Map("+stripe._2.mkString(" , ")+")").saveAsTextFile(args.output())
//    var y=broadcast_var.value(stripe._2._1)
  //  var pmi=Math.log10((total_lines*stripe._2._2)/(x*y))
   // (stripe._1->Map(pmi,stripe._2._1))
    








  }
}
